Add simplified append and of methods to Text

Katrix opened the pull request:
**SpongeAPI** | [SpongeCommon](https://github.com/SpongePowered/SpongeCommon/pull/2161)

Adds `Text#ofCompacted` and `Text#compact` which will try to compact the text when creating it, and when called.

See #1312 for discussion.

Katrix committed:
Add simplified append and of methods to Text

kashike commented on src/main/java/org/spongepowered/api/text/Text.java:
```diff
@@ -535,14 +545,66 @@ public Builder onShiftClick(@Nullable ShiftClickAction<?> shiftClickAction) {
+         *
+         * @return If this builder will simplify when appended text.
+         */
+        public boolean getSimplify() {
```
`shouldSimplify`

kashike commented on src/main/java/org/spongepowered/api/text/Text.java:
```diff
@@ -535,14 +545,66 @@ public Builder onShiftClick(@Nullable ShiftClickAction<?> shiftClickAction) {
+         */
+        public Builder appendSimplified(Text child) {
+            if (!child.isEmpty()) {
+                int lastIndex = this.children.size() - 1;
```
Does this code account for an empty `children` list?

Katrix replied:
Yes next line is a check for if `children` is empty or not. I guess I just forgot to move `lastIndex` into that block.

parlough commented on src/main/java/org/spongepowered/api/text/Text.java:
```diff
@@ -250,6 +250,15 @@ public final boolean isEmpty() {
     /**
+     * Check if this text can be converted to plain and back without loosing any information.
+     *
+     * @return {@code true} if this text is simple.
```
I don't think a code tag is needed here, could just say `@return If this text is simple`

parlough commented on src/main/java/org/spongepowered/api/text/Text.java:
```diff
@@ -250,6 +250,15 @@ public final boolean isEmpty() {
     public abstract Builder toBuilder();
 
     /**
+     * Check if this text can be converted to plain and back without loosing any information.
```
I think this is longer than 80 characters.

parlough commented on src/main/java/org/spongepowered/api/text/Text.java:
```diff
@@ -535,14 +545,66 @@ public Builder onShiftClick(@Nullable ShiftClickAction<?> shiftClickAction) {
         /**
+         * Gets if this builder should try to simplify when appended text.
+         *
+         * @return If this builder will simplify when appended text.
```
No period needed

parlough commented on src/main/java/org/spongepowered/api/text/Text.java:
```diff
@@ -535,14 +545,66 @@ public Builder onShiftClick(@Nullable ShiftClickAction<?> shiftClickAction) {
+        }
+
+        /**
+         * Sets if this builder should simply when appending text.
```
Did you mean "simplify" instead of "simply"?

parlough commented on src/main/java/org/spongepowered/api/text/Text.java:
```diff
@@ -1020,6 +1077,45 @@ public static Text of(Object... objects) {
 
     /**
+     * Builds a simplified {@link Text} from a given array of objects.
+     * This method is identical in every way with {@link #of(Object...)}
```
Put everything starting with "this" in a new paragraph.

parlough reviewed (changes requested):
I suppose I see some use for this. Here's some comments relating to Javadocs.

ryantheleach commented:
On the assumption that this is ready and passes review for the implementation, Does the design as it stands match what was wanted in #1312 ? 

I still think there are some design questions to answer.


Correct me if I'm wrong, but 
As it stands in this PR, it seems that a 'simple' text is only literals, that do not have color, and do not have any mouse actions.

However, it also allows a text to have children that also match this.

In my mind this is not a 'simple' text. I disagree with the design and do not think it matches the expectations of the people in the ticket of #1312.

I see 4 potential designs of simple text.

1. Minimally nested, has the minimum 'depth' of the Text tree to accomplish the format, and hovers the user wants. This means compacting where necessary to flatten, if that makes it shorter (note that this isn't straight forward, due to inheriting default text colors)

2. Text that is just text, whose length can be calculated, and has no hovers or variables like selectors or scoreboards, or translations.

3. Completely flat text. has a single color, and can be split easily, whose length can be calculated.

4. Flat literaltext without color.

I believe 1. is the definition that people in the linked issue wanted.

I don't know whose definition of simple this PR is referring to, or how it is useful to users, I'd like to see proof or a user story for it's inclusion into API7 which it is currently milestoned for.

Katrix commented:
Yeah, I want to rework this into something better myself too. I've since managed to get at least somewhat better implementation than what I have here.

Katrix commented:
So to shed some light on where this is at the moment. I've been sitting on a working implementation for a while now. I cleaned it up some today and added further optimizations. However, I don't feel like I have a good test set to make sure nothing breaks horribly. Anyone mind sending me some busy serialized texts and check that they still look good when displayed?

Here's a current sample of before and after going through the optimizer. As you can see, it will absolutely destroy any form of structure. However, everything I've tried so far still displays correctly.

Before: http://jsoneditoronline.org/?id=1fdee1b7259c4cb51c557ffec11a9a20
After: http://jsoneditoronline.org/?id=c1be09111e2c7d9a8b17f08c5908b4e4

gabizou commented:
Needs to update for Text API v4 changes.

Katrix committed:
Run formatter

JBYoshi commented:
Note: This PR was automatically closed because the `bleeding` branch was deleted. It will need to be reopened to target `api-8` instead.  See https://github.com/SpongePowered/SpongeAPI/pull/1428#issuecomment-539689855.
